lib/Scheduler/enumerator.cpp

```diff
@@ -2040,13 +2040,25 @@ bool LengthCostEnumerator::WasObjctvMet_() {
   InstCount crntCost = GetBestCost_();
 
   InstCount newCost = rgn_->UpdtOptmlSched(crntSched_, this);
-  assert(newCost <= GetBestCost_());
+  if (rgn_->IsSecondPass()) {
+    // for second pass -- newCost is value of RP for sched
+    if (newCost <= rgn_->getFirstPassSpillCost() && 
+        trgtSchedLngth_ <= getBestSchedLength_()) {
+      imprvmntCnt_++;
+    }
 
-  if (newCost < crntCost) {
-    imprvmntCnt_++;
+    return newCost == rgn_->getFirstPassSpillCost();
```

Comment:
One of the preconditions for this code is that the ILP constraint is met -- in other words, the schedule that we have at this given targetLength is a valid schedule. Thus, the only condition we need to check for second pass is if the RP value is the value we found in the first pass. If you look at my changes to updtOptmlSchedule, you'll see that, in the second pass, this function returns the RP value. Thus, we are comparing the second pass RP to first pass RP both to determine optimality and improvement. I'm not thrilled with how updtOptmlSched is coded in this PR, but I couldn't think of another way without introducing updtOptmlSchedScndPss to do second pass specific stuff -- which starts to blur the line between this commit (where weighted cost function is still king) and my next PR wherein everything is split up.

You're right that the condition to increment imprvmntCnt_ is wrong. Essentially, I think it should only happen if (trgtLength < getBestSchedLength() && newCost == getFirstPassSpillCost()) -- that is, if we found a same-RP schedule that is shorter than the one passed to the second pass. I see the confusion about the RP condition here -- in this PR, I decided less than or equal (rather than equal) on the very rare - if not impossible - chance that we find a better RP in the second pass. However, the entire premise of this PR is that we will not find a better RP in the second pass than we found in the first. Either way, though, we shouldn't be updating the firstPassRP - and, on the first encounter of a schedule wherein the second pass RP == firstPassRP, we stop traversing (another premise of this PR) - so I don't think we'll ever be concerned with updating the firstPassSpillCost.